internal/auth/auth.go

```diff
@@ -1,75 +1,129 @@
 package auth
 
 import (
+	"bytes"
 	"context"
+	"crypto/hmac"
+	"crypto/sha1"
+	"encoding/base64"
+	"encoding/hex"
 	"encoding/json"
 	"fmt"
 	"net/http"
+	"strconv"
 	"time"
 )
 
 const (
 	whoAmIPath    = "/apps/users/whoami/"
 	whoamITimeout = 2 * time.Second
+	saltPrefix    = "openslides.utils.sessions.SessionStore"
 )
 
+// Backend gives the sessionData for a sessionID
+type Backend interface {
+	GetSession(sessionID string) ([]byte, error)
+}
+
+// Configer returns the value for a config name.
+type Configer interface {
+	ConfigValue(key string, v interface{}) error
+}
+
 // Auth authentivates a request using the whoami view.
 type Auth struct {
-	addr string
+	cookieName string
+	key        []byte
+	backend    Backend
+	configer   Configer
 }
 
 // New creates a new Auth instance.
-func New(workerAddr string) *Auth {
+func New(cookieName string, secred string, backend Backend, configer Configer) *Auth {
+	bs := sha1.Sum([]byte(saltPrefix + secred))
 	return &Auth{
-		addr: workerAddr,
+		cookieName: cookieName,
+		backend:    backend,
+		configer:   configer,
+		key:        []byte(bs[:]),
 	}
 }
 
-func (a *Auth) whoami(r *http.Request) (int, error) {
-	ctx, close := context.WithTimeout(r.Context(), whoamITimeout)
-	defer close()
+func (a *Auth) validateSessionData(original, value []byte) (bool, error) {
+	mac := hmac.New(sha1.New, a.key)
+	if _, err := mac.Write(value); err != nil {
+		return false, fmt.Errorf("writing hmac")
+	}
 
-	req, err := http.NewRequestWithContext(ctx, http.MethodGet, a.addr+whoAmIPath, nil)
+	originalDecoded := make([]byte, hex.DecodedLen(len(original)))
+	n, err := hex.Decode(originalDecoded, original)
 	if err != nil {
-		return 0, fmt.Errorf("creating whoami request: %w", err)
+		return false, err
 	}
 
-	// Write cookies from the original request to the whoami request.
-	for _, c := range r.Cookies() {
-		req.AddCookie(c)
-	}
+	originalDecoded = originalDecoded[:n]
+	expectedHash := mac.Sum(nil)
+	return hmac.Equal(originalDecoded, expectedHash), nil
+}
 
-	resp, err := http.DefaultClient.Do(req)
+func (a *Auth) userID(r *http.Request) (int, error) {
+	cookie, err := r.Cookie(a.cookieName)
 	if err != nil {
-		return 0, fmt.Errorf("send whoami request: %w", err)
+		if err == http.ErrNoCookie {
+			// Anonymous.
+			return 0, nil
+		}
+		return 0, fmt.Errorf("loading session cookie: %v", err)
 	}
-	defer resp.Body.Close()
 
-	var respData struct {
-		UserID       *int `json:"user_id"`
-		GuestEnabled bool `json:"guest_enabled"`
+	encodedSessionData, err := a.backend.GetSession(cookie.Value)
+	if encodedSessionData == nil {
+		return 0, Error("Unknown session")
```

Comment:
@FinnStutzenstein This error message is also send to the client.

Reply:
Maybe also use `Invalid session data` here.

Reply:
done